File: data/sources/scripts/progress.pl
#!/usr/local/bin/perl
#
# Prints the current progress of the running process, ran by calc_similarity
#
# Author: Itay Maoz, October 2010

use strict;
use File::Basename;
use Cwd;

# constants #
my $CONF_PARSER = "conf_parser.pm";
my $CONF = "conf.txt";
my $TMP_RESULTS_DIR = "tmp_results";
my $STATUS_FILE = "$TMP_RESULTS_DIR/status.txt";

#### Main ####

my $align = shift;
my $num_instances = shift;

usage() unless(defined($align) && defined($num_instances) && $num_instances <= 10);

my $cwd = Cwd::cwd();
my $working_dir = dirname($0);
print "going to $working_dir\n";
chdir($working_dir);

require($CONF_PARSER);

my $conf_file = "../$CONF";

my %conf = parse_configuration($conf_file);
my $results_dir = "";
$results_dir = $conf{"results_dir"};

($results_dir ne "") || die "results dir is not defined in the configuration file";

print "=> results dir = $results_dir\n";

print "=> removing tmp results dir: $TMP_RESULTS_DIR\n";
mysystem("rm -rf $TMP_RESULTS_DIR");

print "=> creating tmp results dir\n";
mkdir($TMP_RESULTS_DIR);

print "=> copying results to the tmp results dir\n";
mysystem("cp $results_dir/*.mat $TMP_RESULTS_DIR");

print "=> getting the progress status\n";
mysystem("./run_matlab_bg.csh 'progress($align,$num_instances);quit' /dev/null");

print "=> waiting for the progress to finish\n";
do {
	print ".\n";
	sleep(1);
} while (!(-f "$STATUS_FILE"));
print "\n";

open (IN, "$STATUS_FILE") || die "can not open status file $STATUS_FILE";
foreach (<IN>) { print $_; }
close(IN);
print "\n";

chdir($cwd);

#### End Main ####

sub mysystem {
	my $cmd = shift;
	print "$cmd\n";
	system("$cmd");
}

sub usage {
        print "Usage:\n";
        print "~~~~~\n\n";
        print "$0 <align> <total number of processes> \n\n";
        print "<align> - 1 to use the aligned descriptors or 0 to use the not aligned descriptors\n";
        print "<total number of processes> - total number of processes that are running on this data (could be on another server)\n";
        print "constraints: <total number of processes> <= 10\n\n";
        exit(-1);
}

File: data/sources/scripts/progress.m
% =========================================================================
%
% Calculate the progress of all the processes that calculate the similarity
% Should be aligned with the progress.pl script (same location and same
% constants).
%
% Inputs:
% ~~~~~~
% align - should run on the aligned version or the not aligned version of 
%         the descriptors
% num_instances - total number of processes to know what splits are in the
%                 responsibility of this instance
%
% Copyright 2010, Lior Wolf, Tal Hassner, and Itay Maoz
%
% =========================================================================

function progress(align, num_instances)

    % constants
    TMP_RESULTS_DIR = './tmp_results/';
    SRC_DIR = '../src';
    CONFIGURATION_FILE = '../conf.txt';
    SIMILARITY_RESULTS_FILE = 'sim_results.mat';

    % validate the configuration file and the source dir exist
    if (~exist(CONFIGURATION_FILE, 'file'))
        error('conf file %s doesnt exist\n', CONFIGURATION_FILE);
        return;
    end

    if (~exist(SRC_DIR, 'dir'))
        error('source dir %s doesnt exist\n', SRC_DIR);
        return;
    end

    addpath(SRC_DIR);

    %load configuration
    conf = load_configuration(CONFIGURATION_FILE)    

    if (~exist('num_instances', 'var'))
        num_instances = -1;
    end   

    results_file = sprintf('%s%s', TMP_RESULTS_DIR, SIMILARITY_RESULTS_FILE);
    
    [sim_results, cnt_per_instance] = unify_similarity_matrix(conf, results_file, align, num_instances);
    num_set = sum(sim_results.splitid > 0);
    total = size(sim_results.splitid, 1);
    time_per_process = int64(sim_results.totaltime / num_instances);
    [hours, minutes, secs] = parse_total_seconds(time_per_process);
    average_time_per_pair = double(time_per_process)/num_set;

    status_file = sprintf('%s%s', TMP_RESULTS_DIR, 'status.txt');
    fid = fopen(status_file,'w');
    fprintf(fid, ['number of calculated similarities: %d out of %d, '... 
                  'average time per process (hh::mm:ss): %.02d:%.02d:%.02d, ',...
                  'average of %f seconds per pair.\n'],...
             num_set, total, hours, minutes, secs, average_time_per_pair);

    percent_done = 100 * num_set/total;
    estimated_secs = int64((total - num_set) * average_time_per_pair);
    [hours, minutes, secs] = parse_total_seconds(estimated_secs);
    fprintf(fid, 'Finished %.2f%%, estimated time to finish (hh::mm:ss): %.02d:%.02d:%.02d\n',...
            percent_done, hours, minutes, secs);

    fprintf(fid, 'count per instance:\n');
    for j=0:(num_instances-1)
        fprintf(fid, 'instance = %d, count = %d\n', j, cnt_per_instance(j+1));
    end
    fclose(fid);
end

function [hours, minutes, seconds] = parse_total_seconds(total_seconds)
	
    seconds = mod(double(total_seconds),60);
    minutes = floor(double(total_seconds)/60);
    minutes = mod(double(minutes),60);
    hours = floor(double(total_seconds)/3600);

end
